resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java

```diff
@@ -1778,33 +1779,56 @@ public void visitAndSave(final IResource root) throws CoreException {
 		// recurse over the projects in the workspace if we were given the workspace root
 		if (root.getType() == IResource.PROJECT)
 			return;
-		IProject[] projects = ((IWorkspaceRoot) root).getProjects(IContainer.INCLUDE_HIDDEN);
+		forEachProjectInParallel(null, this::visitAndSave);
+	}
+
+	@FunctionalInterface
+	public interface CoreConsumer<T> {
+		/**
+		 * Performs this operation on the given argument.
+		 *
+		 * @param t the input argument
+		 */
+		void accept(T t) throws CoreException;
+	}
+
+	private void forEachProjectInParallel(IProgressMonitor m, CoreConsumer<IProject> consumer) throws CoreException {
+		IProject[] projects = workspace.getRoot().getProjects(IContainer.INCLUDE_HIDDEN);
+		if (projects.length == 0) {
+			return;
+		}
+		SubMonitor subMointor = SubMonitor.convert(m, projects.length);
+		IStatus[] stats;
 		// Never use a shared ForkJoinPool.commonPool() as it may be busy with other tasks, which might deadlock.
 		// Also use a custom ForkJoinWorkerThreadFactory, to prevent issues with a
 		// potential SecurityManager, since the threads created by it get no permissions.
 		// See https://github.com/eclipse-platform/eclipse.platform/issues/294
-		ForkJoinPool forkJoinPool = new ForkJoinPool(ForkJoinPool.getCommonPoolParallelism(),
+		ExecutorService executor = new ForkJoinPool(ForkJoinPool.getCommonPoolParallelism(),
 				pool -> new ForkJoinWorkerThread(pool) {
 					// anonymous subclass to access protected constructor
 				}, null, false);
-		IStatus[] stats;
 		try {
-			stats = forkJoinPool.submit(() -> Arrays.stream(projects).parallel().map(project -> {
+			stats = executor.submit(() -> Arrays.stream(projects).parallel().map(project -> {
+				subMointor.split(1);
```

Comment:
Split is usually not useful in a parallel execution, one should use `slice()` instead but need to call done on it after execution.